lib/counters/derive/src/lib.rs

```diff
@@ -0,0 +1,325 @@
+    match gen_count_impl(input) {
+        Ok(tokens) => tokens.to_token_stream().into(),
+        Err(err) => err.to_compile_error().into(),
+    }
+}
+
+fn gen_count_impl(input: DeriveInput) -> Result<impl ToTokens, syn::Error> {
+    let data_enum = match input.data {
+        syn::Data::Enum(ref data_enum) => data_enum,
+        _ => {
+            return Err(syn::Error::new_spanned(
+                input,
+                "`Count` can only be derived for enums",
+            ));
+        }
+    };
+    let variants = &data_enum.variants;
+    let mut state = CountGenerator::new(&input.ident, variants.len());
+
+    for variant in variants {
+        state.add_variant(variant)?;
+    }
+
+    Ok(state.generate(&input))
+}
+
+struct CountGenerator<'input> {
+    enum_name: &'input syn::Ident,
+    field_defs: Vec<proc_macro2::TokenStream>,
+    field_inits: Vec<proc_macro2::TokenStream>,
+    variant_patterns: Vec<proc_macro2::TokenStream>,
+    any_skipped: bool,
+}
+
+impl<'input> CountGenerator<'input> {
+    fn new(enum_name: &'input syn::Ident, variants: usize) -> Self {
+        Self {
+            enum_name,
+            field_defs: Vec::with_capacity(variants),
+            field_inits: Vec::with_capacity(variants),
+            variant_patterns: Vec::with_capacity(variants),
+            any_skipped: false,
+        }
+    }
+
+    fn generate(self, input: &DeriveInput) -> impl ToTokens {
+        let Self {
+            enum_name,
+            field_defs,
+            field_inits,
+            mut variant_patterns,
+            any_skipped,
+        } = self;
+
+        // If we skipped any variants, generate a catchall case.
+        if any_skipped {
+            variant_patterns.push(quote! { _ => {} });
+        }
+        let vis = &input.vis;
+
+        let counts_ty = counts_ty(enum_name);
+        quote! {
+            #[doc = concat!("Total counts for [`", stringify!(#enum_name), "`].")]
+            #[allow(nonstandard_style)]
+            #vis struct #counts_ty {
+                #(#field_defs),*
+            }
+
+            #[automatically_derived]
+            impl counters::Count for #enum_name {
+                type Counters = #counts_ty;
+
+                // This is intended for use in a static initializer, so the fact that every
+                // time the constant is used it will be a different instance is not a
+                // problem --- in fact, it's the desired behavior.
+                //
+                // `declare_interior_mutable_const` is really Not My Favorite Clippy
+                // Lint...
+                #[allow(clippy::declare_interior_mutable_const)]
+                const NEW_COUNTERS: #counts_ty = #counts_ty {
+                    #(#field_inits),*
+                };
+
+                fn count(&self, counters: &Self::Counters) {
+                    #[cfg(all(target_arch = "arm", armv6m))]
+                    use counters::rmv6m_atomic_hack::AtomicU32Ext;
```

Comment:
Typo here?  I'm not sure how it compiles!
```suggestion
                    use counters::armv6m_atomic_hack::AtomicU32Ext;
```

Reply:
Oh, that's weird, huh. Good catch!